### PR TITLE
[OMNI-1506] Add Support for Advanced Jsonschema Features

### PR DESCRIPTION
**Pull request by bc-ninablanson**
# WHAT

Added support for:
### dependencies
- https://github.com/json-schema/json-schema/wiki/Dependencies
- TL:DR; A parent object can state whether any of its child fields depend on other fields.
### oneOf
- https://github.com/json-schema/json-schema/wiki/anyOf,-allOf,-oneOf,-not
- TL:DR; A parent object can specify that it is valid IFF only one of the selected sub-properties are true. 
### Sort based on whether the form an required/object/field hierarchy
- For visual cleanliness, we are sorting the fields in the form based on a hierarchy.
- Required object > Required field > Unrequired object > Unrequired field
# WHY

Need these supports for our forms.
# Checklist
- [ ] I have read and understand the CONTRIBUTIONS.md file
- [ ] I have searched for and linked related issues
- [ ] I have created test cases to ensure quick resolution of the PR is easier
- [ ] I am NOT targeting main branch
- [ ] I did NOT include the dist folder in my PR

### CODE REVIEW

File: src/services/schema-form.js [hunk 1]
```diff
@@ -96,6 +104,109 @@ angular.module('schemaForm').provider('schemaForm',
     return f;
   };
 
+  var combinePropertyDependencies = function(schema) {
```
Comment (bc-ninablanson) [inline src/services/schema-form.js]: I don't quite like how this modifies `properties` - going to figure out a way for this to just modify `items`.

File: dist/schema-form.js [hunk 2]
```diff
@@ -65,6 +65,81 @@ angular.module('schemaForm').provider('sfPath',
   };
```
Comment (bc-ninablanson) [inline dist/schema-form.js:65]: IGNORE - this is the compiled version of all of the changes.

File: src/services/builder.js [hunk 3]
```diff
@@ -113,6 +113,34 @@ angular.module('schemaForm').provider('sfBuilder', ['sfPathProvider', function(s
         }
       }
     },
+    //This is gross, but kind of clean at the same time...
+    oneOfTransclusion: function(args) {
```
Comment (bc-ninablanson) [inline src/services/builder.js:117]: Method that allows one-of values to be transcluded into the DOM.
Sets the value of the ng-show value to be true if the parent object is selected.

File: src/services/schema-form.js [hunk 4]
```diff
@@ -164,6 +275,75 @@ angular.module('schemaForm').provider('schemaForm',
...
+            f.items.push(def);
+          }
+        }
+      });
+
+      return f;
+    }
+  };
+
+  var fieldselect = function(name, schema, options) {
```
Comment (bc-ninablanson) [inline src/services/schema-form.js]: Controls template for oneOf

File: src/services/schema-form.js [hunk 5]
```diff
@@ -409,22 +593,33 @@ angular.module('schemaForm').provider('schemaForm',
...
               global: globalOptions // Global options, including form defaults
             });
+
             if (def) {
               form.push(def);
             }
           }
         });
 
+        if (schema.required && Array === schema.required.constructor) {
```
Comment (bc-ninablanson) [inline src/services/schema-form.js]: Pushes required objects to the top, and then sorts.
This is where the non-required objects can be removed for the initial rollout once all is done.

File: dist/schema-form.js (hunk 2 above)
Comment (rtalvarez) [inline dist/schema-form.js:65]: 💯  had not realized this is under /dist

File: src/services/schema-form.js [hunk 6]
```diff
@@ -4,477 +4,657 @@
...
-
-  // Takes a titleMap in either object or list format and returns one in
-  // in the list format.
-  var canonicalTitleMap = function(titleMap, originalEnum) {
-    if (!angular.isArray(titleMap)) {
-      var canonical = [];
-      if (originalEnum) {
-        angular.forEach(originalEnum, function(value, index) {
-          canonical.push({name: titleMap[value], value: value});
+    ['sfPathProvider', function(sfPathProvider) {
```
Comment (rtalvarez) [inline src/services/schema-form.js:7]: These are just indent changes

File: src/services/schema-form.js [hunk 7]
```diff
@@ -4,477 +4,657 @@
...
-
-          // Do we have form defaults in the schema under the x-schema-form-attribute?
-          if (def.schema['x-schema-form'] && angular.isObject(def.schema['x-schema-form'])) {
-            def = angular.extend(def, def.schema['x-schema-form']);
+        return titleMap;
+      };
+
+      // Takes a titleMap in either object or list format and returns one in
+      // in the list format.
+      var canonicalTitleMap = function(titleMap, originalEnum) {
```
Comment (rtalvarez) [inline src/services/schema-form.js:37]: More indent changes

File: src/services/schema-form.js [hunk 8]
```diff
@@ -4,477 +4,657 @@
...
+        enm.forEach(function(name) {
+          titleMap.push({name: name, value: name});
         });
-      } else {
-        angular.forEach(titleMap, function(name, value) {
-          canonical.push({name: name, value: value});
+        return titleMap;
+      };
+
+      var oneOfToTitleMap = function(ones) {
```
Comment (rtalvarez) [inline src/services/schema-form.js:27]: This oneOfToTitleMap is a legit addition

File: src/services/schema-form.js [hunk 9]
```diff
@@ -4,477 +4,657 @@
...
+
+        // Ng model options doesn't play nice with undefined, might be defined
+        // globally though
+        f.ngModelOptions = f.ngModelOptions || {};
+
+        return f;
+      };
 
-      return postProcessFn(form.map(function(obj) {
+      var combinePropertyDependencies = function(schema) {
```
Comment (rtalvarez) [inline src/services/schema-form.js:107]: this combinePropertyDependencies function is a legit addition

File: src/services/schema-form.js [hunk 10]
```diff
@@ -4,477 +4,657 @@
...
+          });
         }
 
-        if (obj.key) {
-          if (typeof obj.key === 'string') {
-            obj.key = sfPathProvider.parse(obj.key);
+        return schema.properties;
+      };
+
+      var sortRequiredObjectsDesc = function(required, objectFieldNames, form) {
```
Comment (rtalvarez) [inline src/services/schema-form.js:151]: Also a legit addition

File: src/services/schema-form.js [hunk 11]
```diff
@@ -4,477 +4,657 @@
...
+          f.type = 'checkboxes';
+          if (!f.titleMap) {
+            f.titleMap = enumToTitleMap(schema.items['enum']);
+          }
+          options.lookup[sfPathProvider.stringify(options.path)] = f;
+          return f;
         }
+      };
+
+      var row = function(name, schema, options) {
```
Comment (rtalvarez) [inline src/services/schema-form.js:262]: Legit addition

File: src/services/schema-form.js [hunk 12]
```diff
@@ -4,477 +4,657 @@
...
+          if (!isCurrObject && !isCurrRequired) {
+            if (!isNextObject && !isNextRequired) {
+              return 0;
+            }
+            return -1;
+          }
+        })
+      };
+
+      var text = function(name, schema, options) {
```
Comment (rtalvarez) [inline src/services/schema-form.js:194]: Indent changes

File: src/services/schema-form.js [hunk 13]
```diff
@@ -4,477 +4,657 @@
...
+
+          return formObj;
         }
+      };
 
-        //if its has tabs, merge them also!
-        if (obj.tabs) {
-          angular.forEach(obj.tabs, function(tab) {
-            tab.items = service.merge(schema, tab.items, ignore, options, obj.readonly, asyncTemplates);
+      var fieldselect = function(name, schema, options) {
```
Comment (rtalvarez) [inline src/services/schema-form.js:296]: also a legit addition, and pretty much everything down is indent changes
